scispacy/hyponym_detector.py

```diff
@@ -0,0 +1,155 @@
+        self.patterns = BASE_PATTERNS
+        if extended:
+            self.patterns.extend(EXTENDED_PATTERNS)
+
+        self.matcher = Matcher(self.nlp.vocab)
+
+        Doc.set_extension("hearst_patterns", default=[], force=True)
+        Token.set_extension("is_hypernym", default=False, force=True)
+        Token.set_extension("hyponyms", default=[], force=True)
+        Token.set_extension("is_hyponym", default=False, force=True)
+        Token.set_extension("hypernym", default=None, force=True)
+
+        self.predicates = []
+        self.first = set()
+        self.last = set()
+
+        # add patterns to matcher
+        for pattern in self.patterns:
+            self.matcher.add(pattern["label"], None, pattern["pattern"])
+
+            # gather list of predicates where the hypernym appears first
+            if pattern["position"] == "first":
+                self.first.add(pattern["label"])
+
+            # gather list of predicates where the hypernym appears last
+            if pattern["position"] == "last":
+                self.last.add(pattern["label"])
+
+    def expand_to_noun_compound(self, token: Token, doc: Doc):
+        """
+        Expand a token to it's noun phrase based
+        on a simple POS tag heuristic.
+        """
+
+        start = token.i
+        while True:
+            previous = doc[start - 1]
+            if previous.pos_ in {"PROPN", "NOUN", "PRON"}:
+                start -= 1
+            else:
+                break
+
+        end = token.i + 1
+        while True:
+            previous = doc[end]
+            if previous.pos_ in {"PROPN", "NOUN", "PRON"}:
+                end += 1
+            else:
+                break
+
+        return doc[start:end]
+
+    def find_noun_compound_head(self, token: Token):
+
+        while token.head.pos_ in {"PROPN", "NOUN", "PRON"} and token.dep_ == "compound":
+            token = token.head
+        return token
+
+    def __call__(self, doc: Doc):
+        """
+        Runs the matcher on the Doc object and sets token and
+        doc level attributes for hypernym and hyponym relations.
+        """
+        # Find matches in doc
+        matches = self.matcher(doc)
+
+        # If none are found then return None
+        if not matches:
+            return doc
+
+        for match_id, start, end in matches:
+            predicate = self.nlp.vocab.strings[match_id]
+
+            # if the predicate is in the list where the hypernym is last, else hypernym is first
+            if predicate in self.last:
+                hypernym = doc[end - 1]
+                hyponym = doc[start]
+            else:
+                # An inelegent way to deal with the "such_NOUN_as pattern"
+                # since the first token is not the hypernym.
+                if doc[start].lemma_ == "such":
+                    start += 1
+                hypernym = doc[start]
+                hyponym = doc[end - 1]
+
+            hypernym = self.find_noun_compound_head(hypernym)
```

Comment:
are we just appending the head here? shouldn't this be the whole noun chunk?

Reply:
This is adding the token level attributes to only the noun compound's head, but then below use the noun chunk for the document level attribute. I think this is reasonably consistent with spacy's other methods, and also it protects against a token within a noun phrase having a hyponym which is not correct, e.g you could imagine `a country such as Great Britain`, would then get `Great._.hypernym == "country"` which isn't necessarily helpful.

Reply:
hmm, `Britain._.hypernym == "country"` isn't particularly helpful either though, since the full phrase is `Great Britain`. I can't imagine ever wanting to use hypernym/hyponym without getting the whole phrase. Can you? My soft preference would be for hypernym/hyponym attributes to always point to the full noun phrase. The way its currently set up, if I iterate over the doc and check if each token has a hypernym, i can't really get the whole phrase without replicating the noun phrase expansion logic, right? I think it would be better to be able to get the full phrase when iterating over the doc one token at a time.

Reply:
Maybe we should just remove the token level attributes entirely, as I don't think that behaviour is ever desirable/that useful? It's kind of a span level attribute anyway?

Reply:
I can imagine wanting to a know if a specific token participates in a h*nym. If we drop the token level attribute completely, to determine this, you would need to create a `set` out of the token indices (or `dict`) that participate in h*nyms and then check against that `set` (or `dict`). If forcing people to do this themselves sounds fine to you, I'm good with just dropping the token level attributes, but I could imagine having this kind of easy lookup from a token be useful. For example, the lack of this is exactly why writing a spacy matcher for hearst patterns is hard, because tokens don't say if they participate in a noun phrase. Separately, I realized we could actually do this with a matcher pretty easily if we preprocessed to add a token attribute that said whether a token is part of a noun phrase or not, right?

Reply:
EDIT: nvm, you can't match on custom attributes

Reply:
Ok, I think i'll remove them, because it would be weird to have token level attributes for span based properties - spacy doesn't do this (e.g you can't access an NER span from a token). Also it's relatively easy to do this afterward, via span indices.